research/conditional/moe_layers/moe_gating.py

```diff
@@ -0,0 +1,319 @@
+from typing import Union
+import torch
+from fancy_einsum import einsum
+from lizrd.support.logging import make_histogram
+from lizrd.train import checkpointing
+import torch.nn.functional as F
+
+from research.conditional.moe_layers.load_balancing_loss import (
+    calculate_load_balancing_loss,
+)
+from research.conditional.utils.layer_manager import LoggingLayer
+from research.conditional.utils.layer_manager import measure_time
+from lizrd.core.initialization import get_init_fun
+
+
+class MoeGating(LoggingLayer):
+    def __init__(
+        self,
+        n_experts,
+        group_by_batch,
+        softmax_ungrouped,
+        softmax_over,
+        use_torch_bmm,
+        gate,
+    ):
+        super().__init__()
+        self.n_experts = n_experts
+        self.group_by_batch = group_by_batch
+        self.softmax_ungrouped = softmax_ungrouped
+        self.softmax_over = softmax_over
+        self.use_torch_bmm = use_torch_bmm
+        self.gate = gate
+        self._checkpointed_topk_indices: Union[None, torch.Tensor] = None
+
+    def calculate_gate(self, x, batch_size, seq_len):
+        with measure_time(self, "expert_embedding"):
+            if self.use_torch_bmm:
+                gate = self.gate.unsqueeze(0).expand(batch_size, -1, -1)
+                gate_out = torch.bmm(x, gate).permute(2, 0, 1)
+                assert gate_out.shape == (self.n_experts, batch_size, seq_len)
+            else:
+                gate_out = einsum(
+                    "batch_size seq_len dmodel, dmodel n_experts "
+                    "-> n_experts batch_size seq_len ",
+                    x,
+                    self.gate,
+                )
+        # each expert chooses k within dimension 1
+        if not self.group_by_batch and not self.softmax_ungrouped:
+            gate_out = gate_out.reshape(self.n_experts, batch_size * seq_len)
+        # perform softmax either over tokens for each expert or over experts for each token
+        with measure_time(self, "softmax"):
+            if self.softmax_over == "tokens":
+                gate_out = torch.softmax(gate_out, dim=1)
+            elif self.softmax_over == "experts":
+                gate_out = torch.softmax(gate_out, dim=0)
```

Comment:
what if we do typo in `softmax_over`? I believe this assert should be moved/ copied from `token_choice`/`expert_choice` responsibility into here